app/src/main/java/app/security/RequiresPermissionsInterceptor.java

```diff
@@ -0,0 +1,48 @@
+package app.security;
+
+import io.micronaut.aop.InterceptorBean;
+import io.micronaut.aop.MethodInterceptor;
+import io.micronaut.aop.MethodInvocationContext;
+import io.micronaut.core.type.MutableArgumentValue;
+import io.micronaut.http.HttpStatus;
+import io.micronaut.http.exceptions.HttpStatusException;
+import jakarta.inject.Singleton;
+
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Singleton
+@InterceptorBean(RequiresPermissions.class)
+public class RequiresPermissionsInterceptor implements MethodInterceptor<Object, Object> {
```

Comment:
An alternative way of doing this would be to implement the SecurityRule interface and check for the RequiresPermissions annotation.
`Publisher<SecurityRuleResult> check(HttpRequest<?> request, @Nullable RouteMatch<?> routeMatch, @Nullable Authentication authentication);`

The advantage is that the current user  `Authentication` impl is injected and we don't have to do the parsing to access the declaredPermissions.

I don't think it is something that we need to use just mentioning it here as you might be interested.